blpapijs.cpp

```diff
@@ -60,16 +60,38 @@
 #endif
 
 #define BLPAPI_EXCEPTION_TRY try {
-#define BLPAPI_EXCEPTION_CATCH                                              \
-    } catch (blpapi::Exception& e) {                                        \
-        NoRetThrowException(Exception::Error(                               \
-            NEW_STRING(e.description().c_str())));                          \
+
+#define BLPAPI_EXCEPTION_NEW(type)                                          \
+    Local<Object> err = Object::New(Isolate::GetCurrent());                 \
+    err->Set(NEW_STRING("type"), NEW_STRING(#type));                        \
+    err->Set(NEW_STRING("exception"),                                       \
+            Exception::Error(NEW_STRING(e.description().c_str())));
+
+#define BLPAPI_EXCEPTION_THROW(prefix, type)                                \
+    BLPAPI_EXCEPTION_NEW(type)                                              \
+    prefix##RetThrowException(err);
+
+#define BLPAPI_EXCEPTION_CATCH_BLOCK(prefix, type)                          \
+    } catch (blpapi::type& e) {                                             \
+        BLPAPI_EXCEPTION_THROW(prefix, type)
```

Comment:
Would this be a little better if we squashed the above three macros rather than chaining them out? It might be a bit easier to read.